sdk/python/kfp/components/_components.py

```diff
@@ -230,8 +230,8 @@ def create_task_from_component_and_arguments(pythonic_arguments):
     import inspect
     from . import _dynamic
 
-    #Reordering the inputs since in Python optional parameters must come after reuired parameters
-    reordered_input_list = [input for input in inputs_list if not input.optional] + [input for input in inputs_list if input.optional]
+    #Reordering the inputs since in Python optional parameters must come after required parameters
+    reordered_input_list = [input for input in inputs_list if input.default is None and not input.optional] + [input for input in inputs_list if not (input.default is None and not input.optional)]
```

Comment:
In which case, that we will need optional but no default value?

Reply:
Yasser wanted the optional flag.
It's useful in some cases. For instance, when the program expects wither nothing or a proper value, so empty default value does not work.

E.g.
This is valid:
```xgboost --train-data /inputs/train_data --eval-data /inputs/eval_data```
This is valid:
```xgboost --train-data /inputs/train_data```
This is not valid:
```xgboost --train-data /inputs/train_data --eval-data ''```

Reply:
How can you achieve the use case there? Since --eval-data is not part of the inputs, only the value is. This will end up with `xgboost --train-data /inputs/train_data --eval-data` which is probably not valid.

Reply:
This can be handled by the `if` placeholder.
See these two tests:
https://github.com/kubeflow/pipelines/blob/1c4f9eb43178047c2fba68d8495bcec1ee111b63/sdk/python/tests/components/test_components.py#L406
https://github.com/kubeflow/pipelines/blob/1c4f9eb43178047c2fba68d8495bcec1ee111b63/sdk/python/tests/components/test_components.py#L427